### ISSUE TITLE
Optimize diagnostics spectrum calculation performance

### ISSUE THREAD

**Issue by anjor** [anjor/gandalf#52]
## Background
PR #51 code reviews identified critical performance issues in the diagnostics module's spectrum binning implementation. The current Python for-loop prevents JIT compilation and will be extremely slow for production runs at 256³ resolution and above.

### 1. Non-JIT-Compatible Binning Loop (Critical)
**Location**: `src/krmhd/diagnostics.py:149-154`, `diagnostics.py:249-255`

Current implementation:
```python
E_k = jnp.zeros(n_bins)
for i in range(n_bins):
    mask = (k_indices == i)
    bin_energy = jnp.sum(energy_flat * mask)
    E_k = E_k.at[i].set(bin_energy)
```

**Problem**: 
- Python for-loop prevents JIT compilation
- Each iteration creates a new mask and scans the full array
- Very slow for large grids (256³ with 64 bins)

**Solution**: Use `jax.ops.segment_sum` for vectorized binning:
```python
from jax.ops import segment_sum
E_k = segment_sum(energy_flat, k_indices, num_segments=n_bins)
```

**Expected Impact**: 10-100× speedup for spectrum calculations

### 2. Missing @jax.jit Decorators (Moderate)
**Location**: All spectrum functions (`energy_spectrum_1d`, `energy_spectrum_perpendicular`, `energy_spectrum_parallel`)

**Problem**: 
- Per CLAUDE.md: "JIT everything: @jax.jit on all performance-critical functions"
- Spectrum calculations are called every diagnostic interval
- Current implementation recompiles on every call

**Solution**: Add JIT decorators (after fixing the loop issue):
```python
@partial(jax.jit, static_argnames=("n_bins",))
def energy_spectrum_1d(state: KRMHDState, n_bins: int = 32) -> Tuple[Array, Array]:
    # ... implementation
```

**Expected Impact**: 2-5× additional speedup

## Performance Estimates

Based on code review analysis:

| Resolution | Current | With Fixes | Notes |
|-----------|---------|------------|-------|
| 64³       | ~2s     | ~0.5s      | For 100 timesteps + diagnostics |
| 128³      | ~20s    | ~3s        | Bottleneck is binning loop |
| 256³      | ~5min   | ~20s       | JIT + segment_sum critical |

## Tasks
- [ ] Replace Python for-loop with `jax.ops.segment_sum` in all three spectrum functions
- [ ] Add `@partial(jax.jit, static_argnames=("n_bins",))` decorators
- [ ] Verify tests still pass (should be identical output)
- [ ] Add performance benchmark test comparing before/after
- [ ] Update example script with actual measured runtimes

## Acceptance Criteria
- All spectrum functions are JIT-compiled
- No Python for-loops in hot paths
- Tests pass with identical results (within numerical precision)
- Documented 10× speedup on 256³ grids

## Priority
**High** - Critical for production runs at 256³ resolution and above

## Related
- PR #51 (introduced diagnostics module)
- Issue #35 (performance benchmarks for Poisson bracket - similar pattern)
- CLAUDE.md requirement: "JIT everything: @jax.jit on all performance-critical functions"
